### ISSUE TITLE
复现疑问

### ISSUE THREAD

**Issue by mingyao1120** [ssshamiii/Brain-HIVE#1]
感谢作者开源，如下是对应的intra bash
BASE_DIR="-" # TODO
BASE_SEED=2025

# write it here (fixed)
N_REPEATS=3

SUBJECT_IDS=(1 2 3 4 5 6 7 8 9 10)

BRAIN_KEY="eeg"
AVG_TRIALS="true"

SELECTED_CHANNELS='["P7","P5","P3","P1","Pz","P2","P4","P6","P8","PO7","PO3","Oz","PO4","PO8","O1","Oz","O2"]'
PROJ_META='{"RN50":1024,"CLIP-ViT-B-32-laion2B-s34B-b79K":512,"vae":1024}'
TIME_SLICE='[0,250]'

EMB_DIR="${BASE_DIR}/things-eeg/embeddings"
DATA_DIR="${BASE_DIR}/things-eeg/Preprocessed_data_250Hz_whiten"

GPU_CFG="configs/gpu_cfg_single.yaml"
CFG_IN="configs/train_clip.yaml"

for subj in "${SUBJECT_IDS[@]}"; do
  for rep in $(seq 1 "${N_REPEATS}"); do
    # same rep => same seed across subjects
    seed=$((BASE_SEED + rep - 1))
    rep_tag="$(printf "rep-%02d" "${rep}")"

    run_name="RN50+CLIP-ViT-B-32-laion2B-s34B-b79K+vae/intra/subj-${subj}/${rep_tag}"
    output_dir="${BASE_DIR}/exp-${BRAIN_KEY}/${run_name}"
    TMP_FILE="configs/train_clip.${BRAIN_KEY}.resolved.intra.subj-${subj}.${rep_tag}.yaml"

    echo "=== [INTRA] subj=${subj} ${rep_tag}/${N_REPEATS} seed=${seed}"

    python3 build_config.py \
      --config_file "${CFG_IN}" \
      --output_file "${TMP_FILE}" \
      "subject_ids=[${subj},]" \
      "eval_subject_ids=[${subj},]" \
      "brain_key=${BRAIN_KEY}" \
      "run_name=${run_name}" \
      "output_dir=${output_dir}" \
      "avg_trials=${AVG_TRIALS}" \
      "seed=${seed}" \
      "selected_channels=${SELECTED_CHANNELS}" \
      "proj_meta=${PROJ_META}" \
      "time_slice=${TIME_SLICE}" \
      "embedding_directory=${EMB_DIR}" \
      "data_directory=${DATA_DIR}"

    accelerate launch --config_file "${GPU_CFG}" train_clip.py "${TMP_FILE}"
    rm -rf "${TMP_FILE}"
  done
done

以及config和结果，请问为什么得到的性能和论文报告里面差？
{
  "architectures": [
    "BrainCLIPModel"
  ],
  "brain_backbone": "brain_mlp",
  "brain_channels": 17,
  "brain_embedding_dim": 1440,
  "brain_sequence_length": 250,
  "dropout": 0.1,
  "dtype": "float32",
  "hidden_size": 1024,
  "intermediate_size": 1024,
  "logit_scale_init_value": 2.6592,
  "model_type": "clip",
  "proj_meta": {
    "CLIP-ViT-B-32-laion2B-s34B-b79K": 512,
    "RN50": 1024,
    "vae": 1024
  },
  "transformers_version": "5.0.0",
  "use_cache": false
}

{
    "epoch": 25.0,
    "eval_loss": 1.576440453529358,
    "eval_runtime": 0.8777,
    "eval_samples_per_second": 227.877,
    "eval_steps_per_second": 1.139,
    "eval_top1_acc": 0.61,
    "eval_top5_acc": 0.885,
    "total_flos": 1.1784187371225e+17,
    "train_loss": 1.895393824857824,
    "train_runtime": 113.4337,
    "train_samples_per_second": 3645.302,
    "train_steps_per_second": 3.747
}

**Comment by mingyao1120** [comment]
{
    "epoch": 25.0,
    "eval_loss": 3.914473295211792,
    "eval_runtime": 1.2507,
    "eval_samples_per_second": 159.905,
    "eval_steps_per_second": 0.8,
    "eval_top1_acc": 0.095,
    "eval_top5_acc": 0.31
}
您好，我尝试复现inter-subject的训练性能，结果subj1看起来就好像没收敛一样。如下是训练脚本和配置，请有空的时候帮忙看看，谢谢：

```
#!/usr/bin/env bash
set -euo pipefail

BASE_DIR="/data1/-/repository/bci/Brain-HIVE"
BASE_SEED=2025
N_REPEATS=3

ALL_SUBJECTS=(1 2 3 4 5 6 7 8 9 10)

BRAIN_KEY="eeg"
AVG_TRIALS="true"

SELECTED_CHANNELS="['P7', 'P5', 'P3', 'P1','Pz', 'P2', 'P4', 'P6', 'P8', 'PO7', 'PO3', 'POz', 'PO4', 'PO8','O1', 'Oz', 'O2']"
PROJ_META='{"RN50":1024,"CLIP-ViT-B-32-laion2B-s34B-b79K":512,"vae":1024}'
TIME_SLICE='[0,250]'

EMB_DIR="${BASE_DIR}/data/things-eeg/embeddings"
DATA_DIR="${BASE_DIR}/data/things-eeg/Preprocessed_data_250Hz_whiten"

GPU_CFG="configs/gpu_cfg_single.yaml"
CFG_IN="configs/train_clip.yaml"

for eval_subj in "${ALL_SUBJECTS[@]}"; do
  # build train list = ALL \ {eval_subj}
  TRAIN_SUBJECTS=()
  for s in "${ALL_SUBJECTS[@]}"; do
    if [[ "${s}" -ne "${eval_subj}" ]]; then
      TRAIN_SUBJECTS+=("${s}")
    fi
  done
  train_csv="$(IFS=,; echo "${TRAIN_SUBJECTS[*]}")"

  for rep in $(seq 1 "${N_REPEATS}"); do
    seed=$((BASE_SEED + rep - 1))
    rep_tag="$(printf "rep-%02d" "${rep}")"

    run_name="RN50+CLIP-ViT-B-32-laion2B-s34B-b79K+vae/inter/heldout-${eval_subj}/train-[${train_csv}]/${rep_tag}"
    output_dir="${BASE_DIR}/exp-${BRAIN_KEY}/${run_name}"
    TMP_FILE="configs/train_clip.resolved.inter.heldout-${eval_subj}.${rep_tag}.yaml"

    echo "=== [INTER-LOO] heldout=${eval_subj} train=[${train_csv}] ${rep_tag}/${N_REPEATS} seed=${seed}"

    python3 build_config.py \
      --config_file "${CFG_IN}" \
      --output_file "${TMP_FILE}" \
      "subject_ids=[${train_csv},]" \
      "eval_subject_ids=[${eval_subj},]" \
      "brain_key=${BRAIN_KEY}" \
      "run_name=${run_name}" \
      "output_dir=${output_dir}" \
      "avg_trials=${AVG_TRIALS}" \
      "seed=${seed}" \
      "load_best_model_at_end=true" \
      "selected_channels=${SELECTED_CHANNELS}" \
      "proj_meta=${PROJ_META}" \
      "time_slice=${TIME_SLICE}" \
      "embedding_directory=${EMB_DIR}" \
      "data_directory=${DATA_DIR}"

    accelerate launch --config_file "${GPU_CFG}" train_clip.py "${TMP_FILE}"
    rm -rf "${TMP_FILE}"
  done
done


# -----------------------------
# Aggregate retrieval metrics (INTER)
# -----------------------------
EXP_ROOT="${BASE_DIR}/exp-${BRAIN_KEY}/RN50+CLIP-ViT-B-32-laion2B-s34B-b79K+vae/inter"
SUMMARY_JSON="${EXP_ROOT}/retrieval_summary_inter.json"
SUMMARY_CSV="${EXP_ROOT}/retrieval_summary_inter.csv"

echo "=== [AGG] Aggregating eval_results.json under: ${EXP_ROOT}"
python3 build_average.py \
  --exp_root "${EXP_ROOT}" \
  --filename "eval_results.json" \
  --out_json "${SUMMARY_JSON}" \
  --out_csv "${SUMMARY_CSV}"

echo "=== [AGG] Saved:"
echo "  - ${SUMMARY_JSON}"
echo "  - ${SUMMARY_CSV}"
```

```
# Model Arguments

# --- common ---
hidden_size: 1024
logit_scale_init_value: 2.6592
dropout: 0.1

# --- brain ---
brain_backbone: brain_mlp
brain_channels: 17
brain_sequence_length: 250
brain_embedding_dim: 1440

# --- vision ---
proj_meta: ${proj_meta}


# Data Arguments

embedding_directory: null
data_directory: null
time_slice: [0, 250]
subject_ids: [1,]
eval_subject_ids: [1, ]

brain_key: eeg
avg_trials: true
selected_channels: null

cache_dir: .cache

max_train_samples: null
max_eval_samples: null

# Training Arguments
seed: 42
run_name: clip-O+P-H14,B32,VAE
output_dir: null # TODO
resume_from_checkpoint: null

do_train: true
do_eval: true
do_predict: true

per_device_train_batch_size: 1024
per_device_eval_batch_size: 200

learning_rate: 5e-4
weight_decay: 0.05
adam_beta1: 0.9
adam_beta2: 0.995
max_grad_norm: 1.0
num_train_epochs: 25
max_steps: -1
lr_scheduler_type: cosine
warmup_steps: 0

dataloader_num_workers: 8
dataloader_pin_memory: false
remove_unused_columns: false

logging_steps: 50
eval_steps: 0
save_steps: 0
save_strategy: epoch
eval_strategy: epoch
save_total_limit: 1

use_cpu: false
bf16: true
fp16: false
tf32: false
# ddp_backend: nccl

report_to: [swanlab] # [swanlab] # none
load_best_model_at_end: false
metric_for_best_model: top1_acc
greater_is_better: true
eval_do_concat_batches: true

accelerator_config: null
label_names: []
# include_for_metrics: [inputs]

```

> 我们内部复现时结果应该是稳定的，推测应该是pretrained visual encoder 权重不一致问题；稍等我们上传一下我们的 cache 的 visual embedding; 已经 update 到 huggingface，试一下我们 cache 过的 embedding 呢

**Comment by ssshamiii** [comment]
subj-1 的 retrieval top 1 （大概 10% - 15% 上下）和 top 5 accuracy 确实很低，我们报告的也是 10 个 subject 上的均值

**Comment by mingyao1120** [comment]
好的，谢谢，这个算正常波动吗？

**Comment by ssshamiii** [comment]
> 好的，谢谢，这个算正常波动吗？

是的，会发现有时候可以到 15%，有时候只有 9%，所以需要多次 repeat 取均值

**Comment by mingyao1120** [comment]
> > 好的，谢谢，这个算正常波动吗？
> 
> 是的，会发现有时候可以到 15%，有时候只有 9%，所以需要多次 repeat 取均值

好的，感谢您的及时回复！

**Comment by mingyao1120** [comment]
> > 好的，谢谢，这个算正常波动吗？
> 
> 是的，会发现有时候可以到 15%，有时候只有 9%，所以需要多次 repeat 取均值

我这里复现不出来论文的性能。如果您方便的话可以给出Thing-EEG/MEG上inter-subject的逐subj top-1/top-5性能吗？我参考一下

**Comment by ssshamiii** [comment]
> > > 好的，谢谢，这个算正常波动吗？
> > 
> > 
> > 是的，会发现有时候可以到 15%，有时候只有 9%，所以需要多次 repeat 取均值
> 
> 我这里复现不出来论文的性能。如果您方便的话可以给出Thing-EEG/MEG上inter-subject的逐subj top-1/top-5性能吗？我参考一下

如下是我们重新跑的 subj 1 inter 的结果, 感觉您那个有点过于低了

<img width="952" height="759" alt="Image" src="https://github.com/user-attachments/assets/36e97756-6e90-418a-a31f-163eccf070a2" />

**Comment by mingyao1120** [comment]
> > > > 好的，谢谢，这个算正常波动吗？
> > > 
> > > 
> > > 是的，会发现有时候可以到 15%，有时候只有 9%，所以需要多次 repeat 取均值
> > 
> > 
> > 我这里复现不出来论文的性能。如果您方便的话可以给出Thing-EEG/MEG上inter-subject的逐subj top-1/top-5性能吗？我参考一下
> 
> 如下是我们重新跑的 subj 1 inter 的结果, 感觉您那个有点过于低了
> 
> <img alt="Image" width="952" height="759" src="https://private-user-images.githubusercontent.com/87274678/572843339-36e97756-6e90-418a-a31f-163eccf070a2.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NzUwOTcwMDAsIm5iZiI6MTc3NTA5NjcwMCwicGF0aCI6Ii84NzI3NDY3OC81NzI4NDMzMzktMzZlOTc3NTYtNmU5MC00MThhLWEzMWYtMTYzZWNjZjA3MGEyLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNjA0MDIlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjYwNDAyVDAyMjUwMFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWM5MDI2NDAyM2U1ZjA0Y2JhY2NmOGZkNzVmODRhMzk4NWI5YTU2NDRmMjExNzc2NWNiNjllNWM3YzA0ZWNmYmMmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.aII9w3ChmjCchGUE3fO0AeWkcTYlWxRvVca1gujozjA">

好的谢谢您的及时回复，原论文只报告了eeg的平均inter性能，meg这边平均inter和逐subj实验均未给出。因为我想把您的逐subj inter性能作为对比参考，您方便的话可以列在readme文件吗？谢谢！

**Comment by ssshamiii** [comment]
我们当时只保存了平均的结果，并未对单个 subj 的结果进行保存，可以提供其余配置的 inter subj 的结果

<img width="727" height="516" alt="Image" src="https://github.com/user-attachments/assets/a6524b49-7d19-4f48-9861-f49732816ed8" />

**Comment by mingyao1120** [comment]
> 我们当时只保存了平均的结果，并未对单个 subj 的结果进行保存，可以提供其余配置的 inter subj 的结果
> 
> <img alt="Image" width="727" height="516" src="https://private-user-images.githubusercontent.com/87274678/572932665-a6524b49-7d19-4f48-9861-f49732816ed8.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NzUxMjE1MjksIm5iZiI6MTc3NTEyMTIyOSwicGF0aCI6Ii84NzI3NDY3OC81NzI5MzI2NjUtYTY1MjRiNDktN2QxOS00ZjQ4LTk4NjEtZjQ5NzMyODE2ZWQ4LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNjA0MDIlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjYwNDAyVDA5MTM0OVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTRjNjBmNGY3YjVkYmI5YjZlOWYxNzNkNWI5MTdhM2VhNzM4Zjc0Yjc5NThjZTE5NTdmZGYxYzJhYjU2NTEyNDAmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.6OPA3BdLOjtPqhVA-eDnWkG-8v3ONHGcoiOdKDVIevU">

好的，谢谢，我再调整调整
